Test flakes on application tests

## Description

We're seeing occasional failures on the new TBB test logic.

## Steps to Reproduce

1. Inspect CI run output: https://circleci.com/gh/freedomofpress/securedrop/27610
2. Observe failure

## Expected Behavior
Application tests pass reliably. 


## Actual Behavior
Application tests fail occasionally. 

## Comments

The failing tests (observed so far) are:

* `tests/test_2fa.py::test_totp_reuse_protections`
* `tests/test_2fa.py::test_totp_reuse_protections2`
* `tests/test_2fa.py::test_bad_token_fails_to_verify_on_admin_new_user_two_factor_page`
* `tests/test_2fa.py::test_bad_token_fails_to_verify_on_new_user_two_factor_page`
* `tests/test_journalist_api.py::test_user_cannot_get_an_api_token_with_wrong_2fa_token`

Note that none of those are within the `functional/` subdir, but several do emit webdriver-related errors:

```
        elif exception_class == UnexpectedAlertPresentException:
            alert_text = None
            if 'data' in value:
                alert_text = value['data'].get('text')
            elif 'alert' in value:
                alert_text = value['alert'].get('text')
            raise exception_class(message, screen, stacktrace, alert_text)
>       raise exception_class(message, screen, stacktrace)
E       selenium.common.exceptions.WebDriverException: Message: Socket timeout reading Marionette handshake data: Resource temporarily unavailable (os error 11)
```

Also worth noting that the tests are emitting a warning:

```
test_integration.py::test_filenames
  /usr/local/lib/python3.5/dist-packages/flask/sessions.py:208: UserWarning: "localhost" is not a valid cookie domain, it must contain a ".". Add an entry to your hosts file, for example "localhost.localdomain", and use that instead.
    ' "{rv}.localdomain", and use that instead.'.format(rv=rv)
  /usr/local/lib/python3.5/dist-packages/flask/sessions.py:208: UserWarning: "localhost" is not a valid cookie domain, it must contain a ".". Add an entry to your hosts file, for example "localhost.localdomain", and use that instead.
    ' "{rv}.localdomain", and use that instead.'.format(rv=rv)
```

--- a/securedrop/dockerfiles/xenial/python2/Dockerfile
+++ b/securedrop/dockerfiles/xenial/python2/Dockerfile
@@ -30,7 +30,7 @@ RUN curl -LO https://ftp.mozilla.org/pub/firefox/releases/${FF_ESR_VER}/linux-x8
 
 COPY ./tor_project_public.pub /opt/
 
-ENV TBB_VERSION 8.0.8
+ENV TBB_VERSION 8.0.9
 RUN gpg --import /opt/tor_project_public.pub && \
     wget  https://www.torproject.org/dist/torbrowser/${TBB_VERSION}/tor-browser-linux64-${TBB_VERSION}_en-US.tar.xz && \
     wget https://www.torproject.org/dist/torbrowser/${TBB_VERSION}/tor-browser-linux64-${TBB_VERSION}_en-US.tar.xz.asc && \

--- a/securedrop/dockerfiles/xenial/python3/Dockerfile
+++ b/securedrop/dockerfiles/xenial/python3/Dockerfile
@@ -29,7 +29,7 @@ RUN curl -LO https://ftp.mozilla.org/pub/firefox/releases/${FF_ESR_VER}/linux-x8
 
 COPY ./tor_project_public.pub /opt/
 
-ENV TBB_VERSION 8.0.8
+ENV TBB_VERSION 8.0.9
 RUN gpg --import /opt/tor_project_public.pub && \
     wget  https://www.torproject.org/dist/torbrowser/${TBB_VERSION}/tor-browser-linux64-${TBB_VERSION}_en-US.tar.xz && \
     wget https://www.torproject.org/dist/torbrowser/${TBB_VERSION}/tor-browser-linux64-${TBB_VERSION}_en-US.tar.xz.asc && \

--- a/securedrop/tests/conftest.py
+++ b/securedrop/tests/conftest.py
@@ -141,7 +141,7 @@ def alembic_config(config):
 @pytest.fixture(scope='function')
 def source_app(config):
     app = create_source_app(config)
-    app.config['SERVER_NAME'] = 'localhost'
+    app.config['SERVER_NAME'] = 'localhost.localdomain'
     with app.app_context():
         db.create_all()
         yield app
@@ -150,7 +150,7 @@ def source_app(config):
 @pytest.fixture(scope='function')
 def journalist_app(config):
     app = create_journalist_app(config)
-    app.config['SERVER_NAME'] = 'localhost'
+    app.config['SERVER_NAME'] = 'localhost.localdomain'
     with app.app_context():
         db.create_all()
         yield app

--- a/securedrop/tests/functional/functional_test.py
+++ b/securedrop/tests/functional/functional_test.py
@@ -18,6 +18,7 @@ from os.path import realpath
 
 import mock
 import pyotp
+import pytest
 import requests
 import tbselenium.common as cm
 from selenium import webdriver
@@ -40,9 +41,7 @@ from sdconfig import config
 
 os.environ["SECUREDROP_ENV"] = "test"
 
-FUNCTIONAL_TEST_DIR = abspath(dirname(__file__))
-LOGFILE_PATH = abspath(join(FUNCTIONAL_TEST_DIR, "firefox.log"))
-FILES_DIR = abspath(join(dirname(realpath(__file__)), "../..", "tests/files"))
+LOGFILE_PATH = abspath(join(dirname(realpath(__file__)), "../log/driver.log"))
 FIREFOX_PATH = "/usr/bin/firefox/firefox"
 
 TBB_PATH = abspath(join(expanduser("~"), ".local/tbb/tor-browser_en-US/"))
@@ -51,32 +50,20 @@ TBBRC = join(TBB_PATH, "Browser/TorBrowser/Data/Tor/torrc")
 LOGGER.setLevel(logging.WARNING)
 
 
-# https://stackoverflow.com/a/34795883/837471
-class alert_is_not_present(object):
-    """ Expect an alert to not be present."""
-
-    def __call__(self, driver):
-        try:
-            alert = driver.switch_to.alert
-            alert.text
-            return False
-        except NoAlertPresentException:
-            return True
-
-
 class FunctionalTest(object):
-    use_firefox = False
-    driver = None
-    accept_languages = None
-    _firefox_driver = None
-    _torbrowser_driver = None
     gpg = None
-
     new_totp = None
-
+    session_expiration = 30
+    secret_message = "These documents outline a major government invasion of privacy."
     timeout = 10
 
-    secret_message = "These documents outline a major government invasion of privacy."
+    accept_languages = None
+    driver = None
+    firefox_driver = None
+    torbrowser_driver = None
+
+    driver_retry_count = 3
+    driver_retry_interval = 5
 
     def _unused_port(self):
         s = socket.socket()
@@ -85,7 +72,7 @@ class FunctionalTest(object):
         s.close()
         return port
 
-    def _create_torbrowser_driver(self):
+    def create_torbrowser_driver(self):
         logging.info("Creating TorBrowserDriver")
         log_file = open(LOGFILE_PATH, "a")
         log_file.write("\n\n[%s] Running Functional Tests\n" % str(datetime.now()))
@@ -104,125 +91,64 @@ class FunctionalTest(object):
         if self.accept_languages is not None:
             pref_dict["intl.accept_languages"] = self.accept_languages
 
-        self._torbrowser_driver = TorBrowserDriver(
-            TBB_PATH, tor_cfg=cm.USE_RUNNING_TOR, pref_dict=pref_dict, tbb_logfile_path=LOGFILE_PATH
-        )
-        logging.info("Created Tor Browser driver")
+        for i in range(self.driver_retry_count):
+            try:
+                self.torbrowser_driver = TorBrowserDriver(
+                    TBB_PATH,
+                    tor_cfg=cm.USE_RUNNING_TOR,
+                    pref_dict=pref_dict,
+                    tbb_logfile_path=LOGFILE_PATH,
+                )
+                logging.info("Created Tor Browser web driver")
+                break
+            except Exception as e:
+                logging.error("Error creating Tor Browser web driver: %s", e)
+                if i < self.driver_retry_count:
+                    time.sleep(self.driver_retry_interval)
 
-    def _create_firefox_driver(self, profile=None):
-        logging.info("Creating Firefox driver")
-        if profile is None:
-            profile = webdriver.FirefoxProfile()
-            if self.accept_languages is not None:
-                profile.set_preference("intl.accept_languages", self.accept_languages)
-                profile.update_preferences()
+        if not self.torbrowser_driver:
+            raise Exception("Could not create Tor Browser web driver")
 
-        self._firefox_driver = webdriver.Firefox(
-            firefox_binary=FIREFOX_PATH, firefox_profile=profile
-        )
-        self._firefox_driver.set_window_position(0, 0)
-        self._firefox_driver.set_window_size(1024, 768)
-        self._firefox_driver.implicitly_wait(self.timeout)
-        logging.info("Created Firefox driver")
+    def create_firefox_driver(self):
+        logging.info("Creating Firefox web driver")
 
-    def disable_javascript(self):
-        self.driver.profile.set_preference("javascript.enabled", False)
+        profile = webdriver.FirefoxProfile()
+        if self.accept_languages is not None:
+            profile.set_preference("intl.accept_languages", self.accept_languages)
+            profile.update_preferences()
 
-    def enable_javascript(self):
-        self.driver.profile.set_preference("javascript.enabled", True)
+        for i in range(self.driver_retry_count):
+            try:
+                self.firefox_driver = webdriver.Firefox(
+                    firefox_binary=FIREFOX_PATH, firefox_profile=profile
+                )
+                self.firefox_driver.set_window_position(0, 0)
+                self.firefox_driver.set_window_size(1024, 768)
+                self.firefox_driver.implicitly_wait(self.timeout)
+                logging.info("Created Firefox web driver")
+                break
+            except Exception as e:
+                logging.error("Error creating Firefox web driver: %s", e)
+                if i < self.driver_retry_count:
+                    time.sleep(self.driver_retry_interval)
+        if not self.firefox_driver:
+            raise Exception("Could not create Firefox web driver")
 
     def switch_to_firefox_driver(self):
-        self.driver = self._firefox_driver
+        self.driver = self.firefox_driver
+        logging.info("Switched %s to Firefox driver: %s", self, self.driver)
 
     def switch_to_torbrowser_driver(self):
-        self.driver = self._torbrowser_driver
+        self.driver = self.torbrowser_driver
+        logging.info("Switched %s to TorBrowser driver: %s", self, self.driver)
 
-    def setup(self, session_expiration=30):
-        env.create_directories()
-        self.gpg = env.init_gpg()
+    @pytest.fixture(autouse=True)
+    def drivers(self):
+        logging.info("Creating web drivers")
+        self.create_torbrowser_driver()
+        self.create_firefox_driver()
 
-        self.__context = journalist_app.create_app(config).app_context()
-        self.__context.push()
-
-        # Patch the two-factor verification to avoid intermittent errors
-        self.patcher = mock.patch("models.Journalist.verify_token")
-        self.mock_journalist_verify_token = self.patcher.start()
-        self.mock_journalist_verify_token.return_value = True
-
-        self.patcher2 = mock.patch("source_app.main.get_entropy_estimate")
-        self.mock_get_entropy_estimate = self.patcher2.start()
-        self.mock_get_entropy_estimate.return_value = 8192
-
-        signal.signal(signal.SIGUSR1, lambda _, s: traceback.print_stack(s))
-
-        env.create_directories()
-        db.create_all()
-
-        # Add our test user
-        try:
-            valid_password = "correct horse battery staple profanity oil chewy"
-            user = Journalist(username="journalist", password=valid_password, is_admin=True)
-            user.otp_secret = "JHCOGO7VCER3EJ4L"
-            db.session.add(user)
-            db.session.commit()
-        except IntegrityError:
-            logging.error("Test user already added")
-            db.session.rollback()
-
-        # This user is required for our tests cases to login
-        self.admin_user = {
-            "name": "journalist",
-            "password": ("correct horse battery staple" " profanity oil chewy"),
-            "secret": "JHCOGO7VCER3EJ4L",
-        }
-
-        self.admin_user["totp"] = pyotp.TOTP(self.admin_user["secret"])
-
-        source_port = self._unused_port()
-        journalist_port = self._unused_port()
-
-        self.source_location = "http://127.0.0.1:%d" % source_port
-        self.journalist_location = "http://127.0.0.1:%d" % journalist_port
-
-        # Allow custom session expiration lengths
-        self.session_expiration = session_expiration
-
-        self.source_app = source_app.create_app(config)
-        self.journalist_app = journalist_app.create_app(config)
-
-        def start_source_server(app):
-            config.SESSION_EXPIRATION_MINUTES = self.session_expiration
-
-            app.run(port=source_port, debug=True, use_reloader=False, threaded=True)
-
-        def start_journalist_server(app):
-            app.run(port=journalist_port, debug=True, use_reloader=False, threaded=True)
-
-        self.source_process = Process(target=lambda: start_source_server(self.source_app))
-
-        self.journalist_process = Process(
-            target=lambda: start_journalist_server(self.journalist_app)
-        )
-
-        self.source_process.start()
-        self.journalist_process.start()
-
-        for tick in range(30):
-            try:
-                requests.get(self.source_location, timeout=1)
-                requests.get(self.journalist_location, timeout=1)
-            except Exception:
-                time.sleep(0.5)
-            else:
-                break
-
-        self._create_torbrowser_driver()
-        self._create_firefox_driver()
-
-        if self.use_firefox:
-            self.switch_to_firefox_driver()
-        else:
-            self.switch_to_torbrowser_driver()
+        self.switch_to_torbrowser_driver()
 
         # Polls the DOM to wait for elements. To read more about why
         # this is necessary:
@@ -239,6 +165,113 @@ class FunctionalTest(object):
         #
         self.driver.implicitly_wait(self.timeout)
 
+        yield
+
+        try:
+            self.torbrowser_driver.quit()
+        except Exception as e:
+            logging.error("Error stopping TorBrowser driver: %s", e)
+
+        try:
+            self.firefox_driver.quit()
+        except Exception as e:
+            logging.error("Error stopping Firefox driver: %s", e)
+
+    @pytest.fixture(autouse=True)
+    def sd_servers(self):
+        logging.info(
+            "Starting SecureDrop servers (session expiration = %s)", self.session_expiration
+        )
+
+        # Patch the two-factor verification to avoid intermittent errors
+        logging.info("Mocking models.Journalist.verify_token")
+        with mock.patch("models.Journalist.verify_token", return_value=True):
+            logging.info("Mocking source_app.main.get_entropy_estimate")
+            with mock.patch("source_app.main.get_entropy_estimate", return_value=8192):
+
+                try:
+                    signal.signal(signal.SIGUSR1, lambda _, s: traceback.print_stack(s))
+
+                    source_port = self._unused_port()
+                    journalist_port = self._unused_port()
+
+                    self.source_location = "http://127.0.0.1:%d" % source_port
+                    self.journalist_location = "http://127.0.0.1:%d" % journalist_port
+
+                    self.source_app = source_app.create_app(config)
+                    self.journalist_app = journalist_app.create_app(config)
+
+                    self.__context = self.journalist_app.app_context()
+                    self.__context.push()
+
+                    env.create_directories()
+                    db.create_all()
+                    self.gpg = env.init_gpg()
+
+                    # Add our test user
+                    try:
+                        valid_password = "correct horse battery staple profanity oil chewy"
+                        user = Journalist(
+                            username="journalist", password=valid_password, is_admin=True
+                        )
+                        user.otp_secret = "JHCOGO7VCER3EJ4L"
+                        db.session.add(user)
+                        db.session.commit()
+                    except IntegrityError:
+                        logging.error("Test user already added")
+                        db.session.rollback()
+
+                    # This user is required for our tests cases to login
+                    self.admin_user = {
+                        "name": "journalist",
+                        "password": ("correct horse battery staple" " profanity oil chewy"),
+                        "secret": "JHCOGO7VCER3EJ4L",
+                    }
+
+                    self.admin_user["totp"] = pyotp.TOTP(self.admin_user["secret"])
+
+                    def start_source_server(app):
+                        config.SESSION_EXPIRATION_MINUTES = self.session_expiration / 60.0
+
+                        app.run(port=source_port, debug=True, use_reloader=False, threaded=True)
+
+                    def start_journalist_server(app):
+                        app.run(port=journalist_port, debug=True, use_reloader=False, threaded=True)
+
+                    self.source_process = Process(
+                        target=lambda: start_source_server(self.source_app)
+                    )
+
+                    self.journalist_process = Process(
+                        target=lambda: start_journalist_server(self.journalist_app)
+                    )
+
+                    self.source_process.start()
+                    self.journalist_process.start()
+
+                    for tick in range(30):
+                        try:
+                            requests.get(self.source_location, timeout=1)
+                            requests.get(self.journalist_location, timeout=1)
+                        except Exception:
+                            time.sleep(0.25)
+                        else:
+                            break
+                    yield
+                finally:
+                    try:
+                        self.source_process.terminate()
+                    except Exception as e:
+                        logging.error("Error stopping source app: %s", e)
+
+                    try:
+                        self.journalist_process.terminate()
+                    except Exception as e:
+                        logging.error("Error stopping source app: %s", e)
+
+                    env.teardown()
+                    self.__context.pop()
+
     def wait_for_source_key(self, source_name):
         filesystem_id = self.source_app.crypto_util.hash_codename(source_name)
 
@@ -246,17 +279,6 @@ class FunctionalTest(object):
             assert self.source_app.crypto_util.getkey(filesystem_id)
 
         self.wait_for(lambda: key_available(filesystem_id), timeout=60)
-
-    def teardown(self):
-        if self._torbrowser_driver:
-            self._torbrowser_driver.quit()
-        if self._firefox_driver:
-            self._firefox_driver.quit()
-        self.patcher.stop()
-        env.teardown()
-        self.source_process.terminate()
-        self.journalist_process.terminate()
-        self.__context.pop()
 
     def create_new_totp(self, secret):
         self.new_totp = pyotp.TOTP(secret)
@@ -312,13 +334,17 @@ class FunctionalTest(object):
         )
 
     def _alert_accept(self):
+        # adapted from https://stackoverflow.com/a/34795883/837471
+        def alert_is_not_present(object):
+            """ Expect an alert to not be present."""
+            try:
+                alert = self.driver.switch_to.alert
+                alert.text
+                return False
+            except NoAlertPresentException:
+                return True
+
         self.driver.switch_to.alert.accept()
         WebDriverWait(self.driver, self.timeout).until(
-            alert_is_not_present(), "Timed out waiting for confirmation popup to disappear."
-        )
-
-    def _alert_dismiss(self):
-        self.driver.switch_to.alert.dismiss()
-        WebDriverWait(self.driver, self.timeout).until(
-            alert_is_not_present(), "Timed out waiting for confirmation popup to disappear."
+            alert_is_not_present, "Timed out waiting for confirmation popup to disappear."
         )

--- a/securedrop/tests/functional/journalist_navigation_steps.py
+++ b/securedrop/tests/functional/journalist_navigation_steps.py
@@ -96,7 +96,6 @@ class JournalistNavigationStepsMixin:
                 while token == new_token:
                     time.sleep(1)
                     new_token = str(otp.now())
-                    print("Token: {} New token: {}".format(token, new_token))
                 token = new_token
             else:
                 return
@@ -291,7 +290,6 @@ class JournalistNavigationStepsMixin:
 
         if hotp:
             hotp_checkbox = self.driver.find_element_by_css_selector('input[name="is_hotp"]')
-            print(str(hotp_checkbox.__dict__))
             hotp_checkbox.click()
             hotp_secret = self.driver.find_element_by_css_selector('input[name="otp_secret"]')
             hotp_secret.send_keys(hotp)

--- a/securedrop/tests/functional/source_navigation_steps.py
+++ b/securedrop/tests/functional/source_navigation_steps.py
@@ -246,8 +246,8 @@ class SourceNavigationStepsMixin:
     def _source_why_journalist_key(self):
         self.driver.get(self.source_location + "/why-journalist-key")
 
-    def _source_waits_for_session_to_timeout(self, session_length_minutes):
-        time.sleep(session_length_minutes * 60 + 0.1)
+    def _source_waits_for_session_to_timeout(self):
+        time.sleep(self.session_expiration + 2)
 
     def _source_sees_session_timeout_message(self):
         notification = self.driver.find_element_by_css_selector(".important")

--- a/securedrop/tests/functional/test_source_session_timeout.py
+++ b/securedrop/tests/functional/test_source_session_timeout.py
@@ -6,20 +6,12 @@ class TestSourceSessions(
         functional_test.FunctionalTest,
         source_navigation_steps.SourceNavigationStepsMixin):
 
-    def setup(self):
-        # The session expiration here cannot be set to -1
-        # as it will trigger an exception in /create.
-        # Instead, we pick a 1-2s value to allow the account
-        # to be generated.
-        self.session_length_minutes = 0.03
-        super(TestSourceSessions, self).setup(
-            session_expiration=self.session_length_minutes)
+    session_expiration = 5
 
     def test_source_session_timeout(self):
         self._source_visits_source_homepage()
         self._source_clicks_submit_documents_on_homepage()
         self._source_continues_to_submit_page()
-        self._source_waits_for_session_to_timeout(
-            self.session_length_minutes)
-        self._source_visits_source_homepage()
+        self._source_waits_for_session_to_timeout()
+        self.driver.refresh()
         self._source_sees_session_timeout_message()

--- a/securedrop/tests/pageslayout/functional_test.py
+++ b/securedrop/tests/pageslayout/functional_test.py
@@ -39,7 +39,10 @@ class FunctionalTest(functional_test.FunctionalTest):
     def i18n_fixture(self, request):
         logging.debug("i18n_fixture: setting accept_languages to '%s'", request.param)
         self.accept_languages = request.param
-        self.use_firefox = True
+
+    @pytest.fixture(autouse=True)
+    def use_firefox(self):
+        self.switch_to_firefox_driver()
 
     def _screenshot(self, filename):
         log_dir = abspath(

--- a/securedrop/tests/pageslayout/test_source.py
+++ b/securedrop/tests/pageslayout/test_source.py
@@ -131,24 +131,3 @@ class TestSourceLayout(
     def test_why_journalist_key(self):
         self._source_why_journalist_key()
         self._screenshot('source-why_journalist_key.png')
-
-
-@pytest.mark.pagelayout
-class TestSourceSessionLayout(
-        functional_test.FunctionalTest,
-        source_navigation_steps.SourceNavigationStepsMixin,
-        journalist_navigation_steps.JournalistNavigationStepsMixin):
-
-    def setup(self):
-        self.session_length_minutes = 0.03
-        super(TestSourceSessionLayout, self).setup(
-            session_expiration=self.session_length_minutes)
-
-    def test_source_session_timeout(self):
-        self._source_visits_source_homepage()
-        self._source_clicks_submit_documents_on_homepage()
-        self._source_continues_to_submit_page()
-        self._source_waits_for_session_to_timeout(self.session_length_minutes)
-        self._source_enters_text_in_message_field()
-        self._source_visits_source_homepage()
-        self._screenshot('source-session_timeout.png')

--- a/securedrop/tests/pageslayout/test_source.py
+++ b/securedrop/tests/pageslayout/test_source.py
@@ -131,3 +131,21 @@ class TestSourceLayout(
     def test_why_journalist_key(self):
         self._source_why_journalist_key()
         self._screenshot('source-why_journalist_key.png')
+
+
+@pytest.mark.pagelayout
+class TestSourceSessionLayout(
+        functional_test.FunctionalTest,
+        source_navigation_steps.SourceNavigationStepsMixin,
+        journalist_navigation_steps.JournalistNavigationStepsMixin):
+
+    session_expiration = 5
+
+    def test_source_session_timeout(self):
+        self._source_visits_source_homepage()
+        self._source_clicks_submit_documents_on_homepage()
+        self._source_continues_to_submit_page()
+        self._source_waits_for_session_to_timeout()
+        self._source_enters_text_in_message_field()
+        self._source_visits_source_homepage()
+        self._screenshot('source-session_timeout.png')

--- a/securedrop/tests/test_alembic.py
+++ b/securedrop/tests/test_alembic.py
@@ -21,7 +21,7 @@ ALL_MIGRATIONS = [x.split('.')[0].split('_')[0]
                   for x in os.listdir(MIGRATION_PATH)
                   if x.endswith('.py')]
 
-WHITESPACE_REGEX = re.compile('\s*')
+WHITESPACE_REGEX = re.compile(r'\s+')
 
 
 def list_migrations(cfg_path, head):

--- a/securedrop/tests/test_integration.py
+++ b/securedrop/tests/test_integration.py
@@ -1,27 +1,43 @@
 # -*- coding: utf-8 -*-
 
-from distutils.version import StrictVersion
 import gzip
 import os
 import random
 import re
 import zipfile
-import six
-
 from base64 import b32encode
 from binascii import unhexlify
-from bs4 import BeautifulSoup
+from distutils.version import StrictVersion
 from io import BytesIO
-from flask import session, g, escape, current_app
-from pyotp import TOTP, HOTP
+
+import six
+
+import mock
+import pytest
+from bs4 import BeautifulSoup
+from flask import current_app, escape, g, session
+from pyotp import HOTP, TOTP
+
+from . import utils
+from .utils.instrument import InstrumentedApp
 
 os.environ['SECUREDROP_ENV'] = 'test'  # noqa
-from . import utils
 
-from .utils.instrument import InstrumentedApp
 
 # Seed the RNG for deterministic testing
 random.seed('ಠ_ಠ')
+
+
+@pytest.fixture(autouse=True, scope="module")
+def patch_get_entropy_estimate():
+    mock_get_entropy_estimate = mock.patch(
+        "source_app.main.get_entropy_estimate",
+        return_value=8192
+    ).start()
+
+    yield
+
+    mock_get_entropy_estimate.stop()
 
 
 def _login_user(app, user_dict):

--- a/securedrop/tests/test_journalist.py
+++ b/securedrop/tests/test_journalist.py
@@ -160,6 +160,7 @@ def test_unauthorized_access_redirects_to_login(journalist_app):
 
 def test_login_throttle(journalist_app, test_journo):
     # Overwrite the default value used during testing
+    original_hardening = models.LOGIN_HARDENING
     models.LOGIN_HARDENING = True
     try:
         with journalist_app.test_client() as app:
@@ -183,13 +184,14 @@ def test_login_throttle(journalist_app, test_journo):
             assert ("Please wait at least {} seconds".format(
                 Journalist._LOGIN_ATTEMPT_PERIOD) in text)
     finally:
-        models.LOGIN_HARDENING = False
+        models.LOGIN_HARDENING = original_hardening
 
 
 def test_login_throttle_is_not_global(journalist_app, test_journo, test_admin):
     """The login throttling should be per-user, not global. Global login
     throttling can prevent all users logging into the application."""
 
+    original_hardening = models.LOGIN_HARDENING
     # Overwrite the default value used during testing
     # Note that this may break other tests if doing parallel testing
     models.LOGIN_HARDENING = True
@@ -226,7 +228,7 @@ def test_login_throttle_is_not_global(journalist_app, test_journo, test_admin):
             text = resp.data.decode('utf-8')
             assert "Sources" in text
     finally:
-        models.LOGIN_HARDENING = False
+        models.LOGIN_HARDENING = original_hardening
 
 
 def test_login_invalid_credentials(journalist_app, test_journo):
@@ -474,7 +476,7 @@ def test_admin_edits_user_password_error_response(journalist_app,
 def test_user_edits_password_success_response(journalist_app, test_journo):
     original_hardening = models.LOGIN_HARDENING
     try:
-        # Set this to false because we login then immedialtey reuse the same
+        # Set this to false because we login then immediately reuse the same
         # token when authenticating to change the password. This triggers login
         # hardening measures.
         models.LOGIN_HARDENING = False
@@ -499,7 +501,7 @@ def test_user_edits_password_success_response(journalist_app, test_journo):
 def test_user_edits_password_expires_session(journalist_app, test_journo):
     original_hardening = models.LOGIN_HARDENING
     try:
-        # Set this to false because we login then immedialtey reuse the same
+        # Set this to false because we login then immediately reuse the same
         # token when authenticating to change the password. This triggers login
         # hardening measures.
         models.LOGIN_HARDENING = False
@@ -527,7 +529,7 @@ def test_user_edits_password_expires_session(journalist_app, test_journo):
 def test_user_edits_password_error_reponse(journalist_app, test_journo):
     original_hardening = models.LOGIN_HARDENING
     try:
-        # Set this to false because we login then immedialtey reuse the same
+        # Set this to false because we login then immediately reuse the same
         # token when authenticating to change the password. This triggers login
         # hardening measures.
         models.LOGIN_HARDENING = False
@@ -1572,7 +1574,7 @@ def test_render_locales(config, journalist_app, test_journo, test_source):
     # (which are only triggered during `create_app`
     config.SUPPORTED_LOCALES = ['en_US', 'fr_FR']
     app = journalist_app_module.create_app(config)
-    app.config['SERVER_NAME'] = 'localhost'  # needed for url_for
+    app.config['SERVER_NAME'] = 'localhost.localdomain'  # needed for url_for
     url = url_for('col.col', filesystem_id=test_source['filesystem_id'])
 
     # we need the relative URL, not the full url including proto / localhost

--- a/securedrop/tests/test_source.py
+++ b/securedrop/tests/test_source.py
@@ -579,25 +579,26 @@ def test_failed_normalize_timestamps_logs_warning(source_app):
     still occur, but a warning should be logged (this will trigger an
     OSSEC alert)."""
 
-    with patch.object(source_app.logger, 'warning') as logger:
-        with patch.object(subprocess, 'call', return_value=1):
-            with source_app.test_client() as app:
-                new_codename(app, session)
-                _dummy_submission(app)
-                resp = app.post(
-                    url_for('main.submit'),
-                    data=dict(
-                        msg="This is a test.",
-                        fh=(six.StringIO(six.u('')), '')),
-                    follow_redirects=True)
-                assert resp.status_code == 200
-                text = resp.data.decode('utf-8')
-                assert "Thanks! We received your message" in text
+    with patch("source_app.main.get_entropy_estimate", return_value=8192):
+        with patch.object(source_app.logger, 'warning') as logger:
+            with patch.object(subprocess, 'call', return_value=1):
+                with source_app.test_client() as app:
+                    new_codename(app, session)
+                    _dummy_submission(app)
+                    resp = app.post(
+                        url_for('main.submit'),
+                        data=dict(
+                            msg="This is a test.",
+                            fh=(six.StringIO(six.u('')), '')),
+                        follow_redirects=True)
+                    assert resp.status_code == 200
+                    text = resp.data.decode('utf-8')
+                    assert "Thanks! We received your message" in text
 
-                logger.assert_called_once_with(
-                    "Couldn't normalize submission "
-                    "timestamps (touch exited with 1)"
-                )
+                    logger.assert_called_once_with(
+                        "Couldn't normalize submission "
+                        "timestamps (touch exited with 1)"
+                    )
 
 
 def test_source_is_deleted_while_logged_in(source_app):

--- a/securedrop/tests/utils/instrument.py
+++ b/securedrop/tests/utils/instrument.py
@@ -135,7 +135,7 @@ class InstrumentedApp:
         if parts.netloc:
             expected_location = location
         else:
-            server_name = self.app.config.get('SERVER_NAME') or 'localhost'
+            server_name = self.app.config.get('SERVER_NAME') or 'localhost.localdomain'
             expected_location = urljoin("http://%s" % server_name, location)
 
         valid_status_codes = (301, 302, 303, 305, 307)

